src/benchmarklib/tpch/tpch_query_generator.cpp

```diff
@@ -52,6 +70,339 @@ std::string TPCHQueryGenerator::get_preparation_queries() const {
+      const auto diff = date_diff_dist(random_engine);
+      const auto begin_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff * 12);
+      const auto end_date = calculate_date(boost::gregorian::date{1993, 01, 01}, (diff + 1) * 12);
+
+      static std::uniform_real_distribution<> discount_dist{0.02f, 0.09f};
+      const auto discount = discount_dist(random_engine);
+
+      static thread_local std::uniform_int_distribution<> quantity_dist{24, 25};
+      const auto quantity = quantity_dist(random_engine);
+
+      parameters.emplace_back("'"s + begin_date + "'");
+      parameters.emplace_back("'"s + end_date + "'");
+      parameters.emplace_back(std::to_string(discount - 0.01f));
+      parameters.emplace_back(std::to_string(discount + 0.01f));
+      parameters.emplace_back(std::to_string(quantity));
+      break;
+    }
+
+    case 7 - 1: {
+      const auto nation1 = nations.list[nation_dist(random_engine)].text;
+      const auto nation2 = nations.list[nation_dist(random_engine)].text;
+
+      parameters.emplace_back("'"s + nation1 + "'");
+      parameters.emplace_back("'"s + nation2 + "'");
+      parameters.emplace_back("'"s + nation2 + "'");
+      parameters.emplace_back("'"s + nation1 + "'");
+      break;
+    }
+
+    case 8 - 1: {
+      const auto nation_id = nation_dist(random_engine);
+      const auto nation = nations.list[nation_id].text;
+
+      // No idea why the field is called "weight", but it corresponds to the region of a given nation
+      const auto region = regions.list[nations.list[nation_id].weight].text;
+
+      const auto type = p_types_set.list[type_dist(random_engine)].text;
+
+      parameters.emplace_back("'"s + nation + "'");
+      parameters.emplace_back("'"s + region + "'");
+      parameters.emplace_back("'"s + type + "'");
+      break;
+    }
+
+    case 9 - 1: {
+      const auto color = colors.list[color_dist(random_engine)].text;
+
+      parameters.emplace_back("'%"s + color + "%'");
+      break;
+    }
+
+    case 10 - 1: {
+      static thread_local std::uniform_int_distribution<> date_diff_dist{0, 23};
+      const auto diff = date_diff_dist(random_engine);
+      const auto begin_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff);
+      const auto end_date = calculate_date(boost::gregorian::date{1993, 01, 01}, (diff + 3));
+
+      parameters.emplace_back("'"s + begin_date + "'");
+      parameters.emplace_back("'"s + end_date + "'");
+      break;
+    }
+
+    case 11 - 1: {
+      const auto nation = nations.list[nation_dist(random_engine)].text;
+      const auto fraction = 0.0001 / (_scale_factor > 0 ? _scale_factor : 1);
+
+      parameters.emplace_back("'"s + nation + "'");
+      parameters.emplace_back(std::to_string(fraction));
+      parameters.emplace_back("'"s + nation + "'");
+      break;
+    }
+
+    case 12 - 1: {
+      const auto shipmode1 = l_smode_set.list[shipmode_dist(random_engine)].text;
+      std::string shipmode2;
+      do {
```

Comment:
Can we use std::sample to do this more elegantly (for Q7, too, maybe?)

Reply:
> The algorithm is stable (preserves the relative order of the selected elements) only if PopulationIterator meets the requirements of LegacyForwardIterator 

As I am not 100% sure I understand what this does, I don't consider it to be more elegant. People would have to look it up, too.

Reply:
Na, `std::sample` (and `std::shuffle` afterwards) are the algorithms designed to do precisely what is required here. The part you quoted just says that the order of the returned elements is the same as it is in the input range, if the iterators of the input range iterate deterministically (` a == b implies ++a == ++b `), which is not confusing at all.

Reply:
But preserving the order is incorrect. The TPC-H does not define that `shipmode1 < shipmode2`. Admittedly, coming up with a case where this makes a difference in the performance is difficult. Still, I think we should not violate the specification without a good reason (such as me being careless, see above :( )

Reply:
That's way i suggested sample AND shuffle

Reply:
```
      std::array<set_member, 2> random_nations;
      std::sample(&nations.list[0], &nations.list[nations.count], &random_nations[0], 2, random_engine);
      std::shuffle(std::begin(random_nations), std::end(random_nations), random_engine);
```

I really don't consider this more elegant. It's also way slower:

<img width="943" alt="screenshot 2019-01-09 at 13 34 53" src="https://user-images.githubusercontent.com/575106/50899766-63095180-1413-11e9-8ab6-cba071d989c8.png">


http://quick-bench.com/-cjeeb6G2OqFQ583v0LXZzmDqwI